### PR TITLE
fix: Missing plan check and ip whitelist parsing

### LINKED ISSUE #2372: IP Whitelist leading to 500 error in API
### Preliminary Checks

- [X] I have reviewed the documentation: https://unkey.com/docs

- [X] I have searched for existing issues: https://github.com/unkeyed/unkey/issues

- [X] This issue is not a question, general help request, or anything other than a bug report directly related to Unkey. Please ask questions in our Discord community: https://unkey.com/discord.

### Reproduction / Replay Link (Optional)

_No response_

### Issue Summary

When you have the enterprise plan, you have the function to allow IP Whitelisting. This can be done by inputting a list of CSV ips OR newline-separated ips. 

Those IP's are then stored as a csv in the Database.

The issue is that the API tries to JSON parse this, which doesn't work as it's not a JSON string but a CSV, you have the function to allow IP Whitelisting. This can be done by inputting a list of CSV ips OR newline-separated ips

```
2024/10/12 19:18:44 error: {"error":{"code":"INTERNAL_SERVER_ERROR","docs":"https://unkey.dev/docs/api-reference/errors/code/INTERNAL_SERVER_ERROR","message":"Unexpected non-whitespace character after JSON at position 5 (line 1 column 6)","requestId":"req_3UhrGthbDURT4EZhkxZ7oriXsxdQ"}}
```

### Steps to Reproduce

1. Enable the enterprise plan in the db 
2. Visit the API settings and set some IP Whitelist
3. When testing locally be sure to set a IP or comment out the whole part in apps/api/src/pkg/keys/service.ts
CleanShot 2024-10-13 at 10 59 10@2x 

You can then call the verify key method as normal and see that you are getting a 500 error.

### Expected behavior

To avoid getting a 500 error and the IP whitelist to work correctly.

So either the IP Whitelist has to be stored as a JSON array string or the API code has to be adjusted in order to parse csv. 

Question: 
Do the currently non-deleted APIs with IP whitelisting enable have everything stored as CSV or JSON? 
If it's all CSV, it would be easier to adjust the API; otherwise, a script is needed to backfill from JSON to CSV or CSV to JSON. 

### Other information

_No response_

### Screenshots

_No response_

### Version info

```markdown
Irrelevant
```

### PR DESCRIPTION
**Pull request by Flo4604**
## What does this PR do?

This checks whether the Workspace is allowed to have IP whitelisting.
It also fixes the parsing of the whitelisted IPs as which lead to a 500 error before. 

Fixes #2372

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Manually send the updated whitelist topic call with an API and workspace ID with the pro or free plan. This should give an error and not go through.

Afterwards, test by doing a curl and overwriting the True-Client-IP to an invalid and then valid IP  to see if the IP Whitelist works properly.

### Required

- [X] Filled out the "How to test" section in this PR
- [X] Read [Contributing Guide](./CONTRIBUTING.md)
- [X] Self-reviewed my own code
- [X] Commented on my code in hard-to-understand areas
- [X] Ran `pnpm build`
- [X] Ran `pnpm fmt`
- [X] Checked for warnings, there are none
- [X] Removed all `console.logs`
- [X] Merged the latest changes from main onto my branch with `git pull origin main`
- [X] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

## Summary by CodeRabbit

- **New Features**
	- Enhanced input validation for updating IP whitelists, now restricted to enterprise plans.
	- Introduced new request and response schemas for the key verification process, allowing for more detailed interactions.

- **Bug Fixes**
	- Improved handling of `ipWhitelist` property formatting across various tests and services.

- **Documentation**
	- Clarified response type formats for better readability.

- **Tests**
	- Updated test suites to reflect changes in `ipWhitelist` handling and added new tests for rate limit overrides.

### CODE REVIEW

File: apps/dashboard/lib/trpc/routers/api/updateIpWhitelist.ts [hunk 1]
```diff
@@ -43,14 +43,27 @@ export const updateApiIpWhitelist = rateLimitedProcedure(ratelimit.update)
...
+      input.workspaceId !== api.workspace.id
+    ) {
       throw new TRPCError({
         code: "NOT_FOUND",
         message:
           "We are unable to find the correct API. Please try again or contact support@unkey.dev.",
       });
     }
 
+    if (api.workspace.plan !== "enterprise") {
```
Comment (chronark) [inline apps/dashboard/lib/trpc/routers/api/updateIpWhitelist.ts:59]: actually this is a flag on the workspace.features
we decoupled features from plans, cause it's sometimes nice to give someone a feature regardless of their plan

**Review by chronark — CHANGES_REQUESTED** [review]
one issue, rest looks good

File: apps/dashboard/lib/trpc/routers/api/updateIpWhitelist.ts (hunk 1 above)
Comment (Flo4604) [reply apps/dashboard/lib/trpc/routers/api/updateIpWhitelist.ts:59]: Yep makes sense will adjust api and frontend to take that into account

**Comment by Flo4604** [comment]
I will fix the merge conflict.

**Review by chronark — APPROVED** [review]

**Pull request merged**